cf-oscore/src/test/java/org/eclipse/californium/oscore/EndpointContextInfoTest.java

```diff
@@ -0,0 +1,406 @@
+
+	//Keys used in the endpoint context map of strings
+	private final static String OSCORE = OSCoreEndpointContextInfo.OSCORE;
+	private final static String OSCORE_SENDER_ID = OSCoreEndpointContextInfo.OSCORE_SENDER_ID;
+	private final static String OSCORE_RECIPIENT_ID = OSCoreEndpointContextInfo.OSCORE_RECIPIENT_ID;
+	private final static String OSCORE_CONTEXT_ID = OSCoreEndpointContextInfo.OSCORE_CONTEXT_ID;
+	private final static String OSCORE_URI = OSCoreEndpointContextInfo.OSCORE_URI;
+
+	@Before
+	public void initLogger() {
+		System.out.println(System.lineSeparator() + "Start " + getClass().getSimpleName());
+		EndpointManager.clear();
+	}
+
+	//Use the OSCORE stack factory
+	@BeforeClass
+	public static void setStackFactory() {
+		OSCoreCoapStackFactory.useAsDefault(dbClient);
+	}
+
+	@After
+	public void after() {
+		if (null != server) {
+			server.destroy();
+		}
+		System.out.println("End " + getClass().getSimpleName());
+	}
+
+	/**
+	 * Sends two requests and checks that appropriate OSCORE information has
+	 * been set in the following:
+	 * Server: Request source context & Response destination context
+	 * Client: Response source context
+	 *
+	 * @throws Exception if message processing or creating the OSCORE context fails
+	 */
+	@Test
+	public void testEndpointContextInfo() throws Exception {
+	    createSimpleServer();
+
+	    //Set up OSCORE context information for request (client)
+	    byte[] sidClient = new byte[] { 0x77, 0x66, 0x55, 0x44 };
+	    byte[] ridClient = new byte[] { 0x01, 0x02, 0x03, 0x04 };
+	    OSCoreCtx ctx = new OSCoreCtx(master_secret, true, alg, sidClient, ridClient, kdf, 32, null, context_id);
+	    String serverUri = "coap://" + serverEndpoint.getAddress().getAddress().getHostAddress() + ":" + serverEndpoint.getAddress().getPort();
+	    dbClient.addContext(serverUri, ctx);
+
+	    //Create string versions of rid, sid and context ID for client
+	    String sidClientString = DatatypeConverter.printHexBinary(sidClient);
+	    String ridClientString = DatatypeConverter.printHexBinary(ridClient);
+	    String contextIdString = DatatypeConverter.printHexBinary(context_id);
+
+	    String ctxUri = ctx.getUri();
+	    assertNotNull(ctxUri);
+
+	    //Send request
+	    Request request = new Request(CoAP.Code.GET);
+	    request.getOptions().setOscore(Bytes.EMPTY); //Use OSCORE
+	    request.setURI(serverUri);
+	    request.send();
+	    System.out.println("client sent request");
+
+	    //Check original request destination context after transmission
+	    //This is a special case as it should be handled using a callback,
+	    //see testEndpointContextInfoSendingRequest
+	    assertNull(request.getSourceContext());
+	    EndpointContext requestDestinationContext = request.getDestinationContext();
+	    System.out.println("Client: Request destination context type: " + requestDestinationContext.getClass());
+	    assertEquals(AddressEndpointContext.class, requestDestinationContext.getClass());
```

Comment:
If you consider the implementation class of an `EndpointContext` to be significant, it will get much more complexer, than it was intended.

Reply:
And just to mention:
If you modify the destination context in the coap-stack, that's execute by an other thread and the change may not be visible after starting the sending by `request.send();`. 
That "race condition" is one reason, why currently we don't change the destination context.

Reply:
> If you consider the implementation class of an `EndpointContext` to be significant, it will get much more complexer, than it was intended.

I have removed any checks for the explicit class type from the EndpointContextInfoTest. Done in commit 5bbb68aab3d52f095505eb3990090edfcfac71c5

Reply:
> And just to mention:
> If you modify the destination context in the coap-stack, that's execute by an other thread and the change may not be visible after starting the sending by `request.send();`.
> That "race condition" is one reason, why currently we don't change the destination context.

Yes I understand the problem it causes. Changing of destination context has been removed in commit b4dd950c85f774f38aa96f19f57ac9c3e264de2f